//---------------------------------------------------------------------------
//
// DLS protocol definition
//
// vim: syntax=proto tw=78
//
//---------------------------------------------------------------------------

package DlsProto;

//---------------------------------------------------------------------------

message Hello {
    required string version = 1;
    required string revision = 2;
    required int32 protocol_version = 3;
}

//---------------------------------------------------------------------------

message Request {
    optional DirInfoRequest dir_info = 1;
    optional JobRequest job_request = 2;
}

message DirInfoRequest {
    optional string path = 1;
}

message JobRequest {
    required uint32 id = 1;
    optional bool fetch_channels = 2;
    optional ChannelRequest channel_request = 3;
}

message ChannelRequest {
    required uint32 id = 1;
    optional bool fetch_chunks = 2;
}

//---------------------------------------------------------------------------

message Response {
    optional Error error = 1;
    optional DirInfo dir_info = 2;
}

message DirInfo {
    optional string path = 1;
    repeated JobInfo job = 2;
}

message JobInfo {
    optional JobPresetInfo preset = 1;
    repeated ChannelInfo channel = 2;
}

message JobPresetInfo {
    required uint32 id = 1;
    optional string description = 2;
    optional bool running = 3;
    optional uint64 quota_time = 4;
    optional uint64 quota_size = 5;
    optional string source = 6;
    optional uint32 port = 7; // no uint16 available
    optional string trigger = 8;
    // FIXME channel presets
}

message ChannelInfo {
    required uint32 id = 1;
    optional string name = 2;
    optional string unit = 3;
    optional ChannelType type = 4;
    optional bool recording = 5;
    repeated ChunkInfo chunk = 6;
}

message ChunkInfo {
    required uint64 start = 4;
    required uint64 end = 5;
}

message Error {
    optional string message = 1;
}

//---------------------------------------------------------------------------

enum ChannelType
{
    TUNKNOWN = 0;
    TCHAR = 1;
    TUCHAR = 2;
    TSHORT = 3;
    TUSHORT = 4;
    TINT = 5;
    TUINT = 6;
    TLINT = 7;
    TULINT = 8;
    TFLT = 9;
    TDBL = 10;
}

//---------------------------------------------------------------------------
